dataplane/forwarding/fwdbuilder/action.go

```diff
@@ -0,0 +1,201 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package fwdbuilder contains builders for varius proto types.
+package fwdbuilder
+
+import (
+	"encoding/binary"
+
+	fwdpb "github.com/openconfig/lemming/proto/forwarding"
+)
+
+type actionDescBuilder interface {
+	set(*fwdpb.ActionDesc)
+	atype() fwdpb.ActionType
+}
+
+var (
+	// Compile-time checks that builders implement actionDescBuilder interface.
+	_ actionDescBuilder = &LookupActionBuilder{}
+	_ actionDescBuilder = &TransmitActionBuilder{}
+	_ actionDescBuilder = &LookupActionBuilder{}
+)
+
+// ActionBuilder is a builder for forward action types.
+type ActionBuilder struct {
+	adb actionDescBuilder
+}
+
+// Action returns a new action builder.
+func Action(adb actionDescBuilder) *ActionBuilder {
+	return &ActionBuilder{
+		adb: adb,
+	}
+}
+
+// WithActionDesc sets the action description
+func (ab *ActionBuilder) WithActionDesc(adb actionDescBuilder) *ActionBuilder {
+	ab.adb = adb
+	return ab
+}
+
+// Build returns a new action.
+func (ab *ActionBuilder) Build() *fwdpb.ActionDesc {
+	a := &fwdpb.ActionDesc{
+		ActionType: ab.adb.atype(),
+	}
+	ab.adb.set(a)
+	return a
+}
+
+// UpdateActionBuilder is a builder for an update action.
+type UpdateActionBuilder struct {
+	fieldIDNum fwdpb.PacketFieldNum
+	updateType fwdpb.UpdateType
+	fieldSrc   fwdpb.PacketFieldNum
```

Comment:
Is this the right type? What does this field represent?

Reply:
yes, for COPY update type src header